From 797a0cbe691dbcc173e7489db3b9f23f0747749c Mon Sep 17 00:00:00 2001
From: Kiet Pham <kiet.pham.xb@renesas.com>
Date: Thu, 29 Aug 2024 11:00:54 +0700
Subject: [PATCH 3/4] omxvideodec: Add enable-scale option

From OMX v1.3.0, non-bypass mode supports scaling as post processing. So,
this option is added for user to scale output image via output caps.
This feature currently not support if input resolution is smaller than or
equal to 96x80.

Signed-off-by: Kiet Pham <kiet.pham.xb@renesas.com>
---
 omx/gstomxvideodec.c | 163 ++++++++++++++++++++++++++++++++++++++++++-
 omx/gstomxvideodec.h |   2 +
 2 files changed, 164 insertions(+), 1 deletion(-)

diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 70a4bfa..d56c74d 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -92,6 +92,12 @@ static gboolean gst_omx_video_dec_deallocate_output_buffers (GstOMXVideoDec

 static gboolean gst_omx_video_dec_get_cropped_resolution (GstOMXVideoDec * self,
     gint * out_width, gint * out_height);
+static gboolean gst_omx_video_dec_get_resolution_from_src_pad (GstOMXVideoDec * self,
+    gint * out_width, gint * out_height);
+static gboolean get_omx_video_dec_set_scale (GstOMXVideoDec * self,
+    gint in_width, gint in_height);
+static gint gst_omx_video_dec_calculate_scale_ratio (gint new_length,
+    gint length);

 enum
 {
@@ -103,7 +109,8 @@ enum
   PROP_LOSSY_COMPRESS,
   PROP_ENABLE_CROP,
   PROP_BYPASS,
-  PROP_NUM_OUTPUT_BUFFER
+  PROP_NUM_OUTPUT_BUFFER,
+  PROP_ENABLE_SCALE,
 };

 #define GST_OMX_VIDEO_DEC_INTERNAL_ENTROPY_BUFFERS_DEFAULT (5)
@@ -113,6 +120,8 @@ enum
 #define GST_OMX_VIDEO_DEC_MIN_FRAMEHEIGHT                  (80)
 #define GST_OMX_VIDEO_DEC_MIN_STRIDE                       (96)
 #define GST_OMX_VIDEO_DEC_MIN_SLICEHEIGHT                  (80)
+#define GST_OMX_VIDEO_DEC_MAX_FRAMEWIDTH                   (1920)
+#define GST_OMX_VIDEO_DEC_MAX_FRAMEHEIGHT                  (1080)

 /* class initialization */

@@ -159,10 +168,18 @@ gst_omx_video_dec_set_property (GObject * object, guint prop_id,
 #endif
     case PROP_BYPASS:
       self->bypass = g_value_get_boolean (value);
+      if (self->bypass == TRUE) {
+        self->enable_scale = FALSE;
+      }
       break;
     case PROP_NUM_OUTPUT_BUFFER:
       self->num_outbufs = g_value_get_uint (value);
       break;
+    case PROP_ENABLE_SCALE:
+      self->enable_scale = g_value_get_boolean (value);
+      if (self->bypass == TRUE)
+        self->enable_scale = FALSE;
+      break;
 #else
     case PROP_NO_REORDER:
       GST_WARNING_OBJECT (self, "HAVE_VIDEODEC_EXT not enabled. Couldn't configure property no-reorder (require vendor specific implement).\n");
@@ -213,6 +230,9 @@ gst_omx_video_dec_get_property (GObject * object, guint prop_id,
     case PROP_NUM_OUTPUT_BUFFER:
       g_value_set_uint (value, self->num_outbufs);
       break;
+    case PROP_ENABLE_SCALE:
+      g_value_set_boolean (value, self->enable_scale);
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -283,6 +303,12 @@ gst_omx_video_dec_class_init (GstOMXVideoDecClass * klass)
           GST_OMX_VIDEO_DEC_NUMBER_OUTPUT_BUFFERS_DEFAULT,
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
           GST_PARAM_MUTABLE_READY));
+  g_object_class_install_property (gobject_class, PROP_ENABLE_SCALE,
+      g_param_spec_boolean ("enable-scale",
+          "Scaling video using src caps ",
+          "Whether or not to enable scaling if Bypass mode is disabled",
+          TRUE, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
+          GST_PARAM_MUTABLE_READY));

   element_class->change_state =
       GST_DEBUG_FUNCPTR (gst_omx_video_dec_change_state);
@@ -328,6 +354,7 @@ gst_omx_video_dec_init (GstOMXVideoDec * self)
   self->lossy_compress = FALSE;
   self->has_set_property = FALSE;
   self->bypass = FALSE;
+  self->enable_scale = FALSE;
   /* The default value is 0, which means the number of output buffers will be
    * automatically updated when allocated */
   self->num_outbufs = GST_OMX_VIDEO_DEC_NUMBER_OUTPUT_BUFFERS_DEFAULT;
@@ -1614,6 +1641,35 @@ gst_omx_video_dec_reconfigure_output_port (GstOMXVideoDec * self)
     goto done;
   }

+  /* FIXME: In case decode FullHD video, the decode size is 1920x1088. However,
+   * maxmimum supported image size of G2L is only 1920x1080. So, output buffer
+   * could not be used by downstream plugins */
+  port_def.format.video.nFrameHeight =
+      MIN (port_def.format.video.nFrameHeight, GST_OMX_VIDEO_DEC_MAX_FRAMEHEIGHT);
+
+  /* Update scale ratio base on decoded information */
+  if (self->enable_scale == TRUE) {
+    gint cropped_width, cropped_height;
+
+    if (!gst_omx_video_dec_get_cropped_resolution (self, &cropped_width,
+                                                   &cropped_height))
+      goto done;
+
+    if (!get_omx_video_dec_set_scale (self, cropped_width, cropped_height))
+      goto done;
+
+    /* If there is cropped information in SPS, it means scale ratio is
+     * calculated incorrectly. So, nStride / nSliceHeight updated by OMX should
+     * not be used. */
+    if (!gst_omx_video_dec_get_resolution_from_src_pad (self, &cropped_width,
+                                                        &cropped_height))
+      goto done;
+    port_def.format.video.nStride      = cropped_width;
+    port_def.format.video.nSliceHeight = cropped_height;
+  }
+
+  gst_omx_port_update_port_definition (port, &port_def);
+
   GST_DEBUG_OBJECT (self,
       "Setting output state: format %s (%d), width %u, height %u",
       gst_video_format_to_string (format),
@@ -1836,6 +1892,75 @@ gst_omx_video_dec_get_cropped_resolution (GstOMXVideoDec * self,
   return TRUE;
 }

+static gboolean
+gst_omx_video_dec_get_resolution_from_src_pad (GstOMXVideoDec * self,
+    gint * out_width, gint * out_height)
+{
+  GstCaps *templ_caps, *intersection;
+  GstStructure *s;
+
+  templ_caps = gst_pad_get_pad_template_caps (GST_VIDEO_DECODER_SRC_PAD (self));
+  intersection =
+      gst_pad_peer_query_caps (GST_VIDEO_DECODER_SRC_PAD (self), templ_caps);
+  gst_caps_unref (templ_caps);
+
+  GST_DEBUG_OBJECT (self, "Allowed downstream caps: %" GST_PTR_FORMAT,
+      intersection);
+
+  s = gst_caps_get_structure (intersection, 0);
+  if (gst_structure_has_field (s, "width")) {
+    gst_structure_get_int (s, "width", out_width);
+  }
+  if (gst_structure_has_field (s, "height")) {
+    gst_structure_get_int (s, "height", out_height);
+  }
+
+  gst_caps_unref (intersection);
+  return TRUE;
+}
+
+static gboolean
+get_omx_video_dec_set_scale (GstOMXVideoDec * self,
+    gint in_width, gint in_height)
+{
+  gint out_width = in_width, out_height = in_height;
+
+  if (!gst_omx_video_dec_get_resolution_from_src_pad (self,
+                                                      &out_width, &out_height))
+    return FALSE;
+
+  if (out_width % 2 || out_height % 2) {
+      GST_ERROR_OBJECT (self, "Unsupported resolution: %dx%d",
+                        out_width, out_height);
+      return FALSE;
+  }
+
+  if (out_width <= in_width && out_height <= in_height) {
+    OMX_CONFIG_SCALEFACTORTYPE sScale;
+    OMX_ERRORTYPE err;
+    GST_OMX_INIT_STRUCT (&sScale);
+    sScale.nPortIndex = self->dec_out_port->index;
+    sScale.xWidth =
+        gst_omx_video_dec_calculate_scale_ratio (out_width, in_width);
+    sScale.xHeight =
+        gst_omx_video_dec_calculate_scale_ratio (out_height, in_height);
+
+    err = gst_omx_component_set_config (self->dec,
+                                        OMX_IndexConfigCommonScale, &sScale);
+    if (err != OMX_ErrorNone) {
+      GST_ERROR_OBJECT (self, "Failed to update scale propety: %s (0x%08x)",
+                        gst_omx_error_to_string (err), err);
+      return FALSE;
+    }
+  } else {
+    GST_ERROR_OBJECT (self, "Unsupported scale up: %dx%d to %dx%d",
+                      in_width, in_height, out_width, out_height);
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
 static void
 gst_omx_video_dec_update_buffer_meta (GstOMXVideoDec * self, GstBuffer * buffer,
     gint out_width, gint out_height, GstVideoMeta * vmeta)
@@ -2056,6 +2181,12 @@ gst_omx_video_dec_loop (GstOMXVideoDec * self)
     if (!gst_omx_video_dec_get_cropped_resolution (self, &out_width, &out_height))
       goto component_error;

+    if (self->enable_scale == TRUE) {
+      if (!gst_omx_video_dec_get_resolution_from_src_pad (self, &out_width,
+                                                          &out_height))
+        goto component_error;
+    }
+
     if (!gst_omx_video_dec_update_output_state (self, out_width, out_height))
       goto caps_failed;
   }
@@ -3056,6 +3187,20 @@ gst_omx_video_dec_set_format (GstVideoDecoder * decoder,
       self->bypass = sBypass.bEnable;
     }
   }
+
+  if (!self->disabled && self->enable_scale == TRUE) {
+    /* Setting scale ratio base on current video information. If resolution is
+     * smaller than or equal to 96x80, there is no PSC returned by OMX.
+     * Therefore, scaling is not supported for these resolution */
+    if (info->width > 96 || info->height > 80) {
+      if (!get_omx_video_dec_set_scale (self, info->width, info->height))
+        return FALSE;
+    } else {
+      GST_ERROR_OBJECT (self, "Unsupported scaling for %dx%d resolution",
+                        info->width, info->height);
+      return FALSE;
+    }
+  }
 #else
   if (self->no_reorder != FALSE)
     GST_ERROR_OBJECT (self,
@@ -3813,3 +3958,19 @@ gst_omx_video_dec_propose_allocation (GstVideoDecoder * bdec, GstQuery * query)
       GST_VIDEO_DECODER_CLASS
       (gst_omx_video_dec_parent_class)->propose_allocation (bdec, query);
 }
+
+static gint
+gst_omx_video_dec_calculate_scale_ratio (gint new_length, gint length)
+{
+  gulong scale_ratio;
+
+  if (new_length == 0 || length == 0) {
+    return 0x10000;
+  }
+
+  scale_ratio = ((gulong)new_length << 32) / length;
+
+  /* Convert to Q15.16 format */
+  return (gint)(scale_ratio >> 16);
+}
+
diff --git a/omx/gstomxvideodec.h b/omx/gstomxvideodec.h
index f8d79c7..6fc8e9c 100644
--- a/omx/gstomxvideodec.h
+++ b/omx/gstomxvideodec.h
@@ -116,6 +116,8 @@ struct _GstOMXVideoDec
   gboolean bypass;
   /* Number of output buffers that are required on output port */
   guint32 num_outbufs;
+  /* Set TRUE to scale output video using src caps */
+  gboolean enable_scale;
 };

 struct _GstOMXVideoDecClass
--
2.25.1

